### PR TITLE
Expand contact tests

### PR DESCRIPTION
**Pull request by mglaman**
Expands functional tests to test contact fields and contact creation.

Currently a draft due to pending fixups.

### CODE REVIEW

File: tests/src/FunctionalJavascript/ContactSubmissionTest.php [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+   * During tests configuration schema is validated. This module does not
+   * provide schema definitions for its handler.
+   *
+   * To fix: webform.webform.civicrm_webform_test:handlers.webform_civicrm.settings
+   *
+   * @see \Drupal\Core\Test\TestSetupTrait::getConfigSchemaExclusions
+   */
+  protected static $configSchemaCheckerExclusions = [
+    'webform.webform.civicrm_webform_test',
+  ];
```
Comment (mglaman) [inline tests/src/FunctionalJavascript/ContactSubmissionTest.php]: this is in the parent class

File: tests/src/FunctionalJavascript/ContactSubmissionTest.php [hunk 2]
```diff
@@ -0,0 +1,196 @@
+<?php
+
+namespace Drupal\Tests\webform_civicrm\FunctionalJavascript;
+
+use Behat\Mink\Exception\ElementNotFoundException;
```
Comment (mglaman) [inline tests/src/FunctionalJavascript/ContactSubmissionTest.php]: Unused import

File: tests/src/FunctionalJavascript/ContactSubmissionTest.php [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+   *
+   * @dataProvider dataContactValues
+   */
+  public function testSubmitWebform($contact_type, array $contact_values) {
+    $this->assertArrayHasKey('contact', $contact_values, 'Test data must contain contact');
+    $this->assertArrayHasKey('first_name', $contact_values['contact'], 'Test contact data must contain first_name');
+    $this->assertArrayHasKey('last_name', $contact_values['contact'], 'Test confact data must contain last_name');
+
+    // @todo change to login as admin then logout, then login as test user for submission
+    $this->drupalLogin($this->testUser);
```
Comment (mglaman) [inline tests/src/FunctionalJavascript/ContactSubmissionTest.php]: Currently `testUser` is an admin, and should be renamed as such. Then testUser should be a basic authenticated (or anonymous?) with permissions to submit the form.

File: tests/src/FunctionalJavascript/ContactSubmissionTest.php [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+      $this->assertSession()->checkboxChecked('civicrm_1_contact_1_email_email');
+    }
+    if (isset($contact_values['website'])) {
+      $this->assertTrue(is_array($contact_values['website']));
+      $this->assertTrue(isset($contact_values['website'][0]));
+      $this->getSession()->getPage()->selectFieldOption('contact_1_number_of_website', count($contact_values['website'][0]));
+      $this->assertSession()->assertWaitOnAjaxRequest();
+      $this->htmlOutput();
+      $this->assertSession()->checkboxChecked('civicrm_1_contact_1_website_url');
+    }
```
Comment (mglaman) [inline tests/src/FunctionalJavascript/ContactSubmissionTest.php]: This is a lot of duplicate code, especially when Phone, Messaging, and Address are added

File: tests/src/FunctionalJavascript/ContactSubmissionTest.php [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        else {
+          $selector = "civicrm_1_contact_1_{$entity_type}_{$field_name}";
+          $this->getSession()->getPage()->fillField($selector, $field_value);
+        }
+      }
+    }
+    $this->getSession()->getPage()->pressButton('Submit');
+    $this->assertSession()->pageTextContains('New submission added to CiviCRM Webform Test.');
+    // @todo replace with wf_civicrm_api()
+    $contact_result = civicrm_api('contact', 'get', [
```
Comment (mglaman) [inline tests/src/FunctionalJavascript/ContactSubmissionTest.php]: Use wf_civicrm_api

File: tests/src/FunctionalJavascript/ContactSubmissionTest.php [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    $this->assertEquals($contact_type, $contact['contact_type']);
+
+    foreach ($contact_values['contact'] as $field_name => $field_value) {
+      $this->assertEquals($field_value, $contact[$field_name], $result_debug);
+    }
+
+    if (isset($contact_values['email'])) {
+      $this->assertEquals($contact_values['email'][0]['email'], $contact['email']);
+      // @todo replace with wf_civicrm_api()
+      $email_result = civicrm_api('email', 'get', [
```
Comment (mglaman) [inline tests/src/FunctionalJavascript/ContactSubmissionTest.php]: Use wf_civicrm_api

File: tests/src/FunctionalJavascript/ContactSubmissionTest.php [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        'contact_id' => $contact['contact_id'],
+      ]);
+      $this->assertEquals(count($contact_values['email']), $email_result['count']);
+      foreach ($email_result['values'] as $key => $email_entity) {
+        $this->assertEquals($contact_values['email'][$key]['email'], $email_entity['email']);
+      }
+    }
+    if (isset($contact_values['website'])) {
+      // @todo replace with wf_civicrm_api()
+      $website_result = civicrm_api('website', 'get', [
```
Comment (mglaman) [inline tests/src/FunctionalJavascript/ContactSubmissionTest.php]: Use wf_civicrm_api

**Review by mglaman — COMMENTED** [review]
Some feedback for myself

**Comment by mglaman** [comment]
@KarinG okay, this tests Phone, Email, Website. I didn't add IM (forgot, also I figure it's a lesser example?) or Address (probably a bigger effort and follow up.)

This adds some good expanded contact testing. Next would be address and the different "types" of phone/email/website.

I think I might have some naming wrong. These are called Location types? Even though it is not a physical location (looking at `wf_crm_location_fields`(

**Comment by KarinG** [comment]
Awesome! All checks have passed :-)

**Pull request merged**